config-controller/pkg/client/client.go

```diff
@@ -0,0 +1,203 @@
+	Create(ctx context.Context, policy *storage.Policy) (*storage.Policy, error)
+	Update(ctx context.Context, policy *storage.Policy) error
+	FlushCache(ctx context.Context) error
+	EnsureFresh(ctx context.Context) error
+}
+
+//go:generate mockgen-wrapper
+type PolicyClient interface {
+	List(context.Context) ([]*storage.ListPolicy, error)
+	Get(ctx context.Context, id string) (*storage.Policy, error)
+	Post(context.Context, *storage.Policy) (*storage.Policy, error)
+	Put(context.Context, *storage.Policy) error
+}
+
+type grpcClient struct {
+	svc v1.PolicyServiceClient
+}
+
+func newGrpcClient(ctx context.Context) (PolicyClient, error) {
+	conn, err := common.GetGRPCConnection(auth.TokenAuth(), logger.NewLogger(roxctlIO.DefaultIO(), printer.DefaultColorPrinter()))
+	if err != nil {
+		return nil, errors.Wrap(err, "could not establish gRPC connection to Central")
+	}
+	svc := v1.NewPolicyServiceClient(conn)
+
+	return &grpcClient{
+		svc: svc,
+	}, nil
+}
+
+func (gc *grpcClient) List(ctx context.Context) ([]*storage.ListPolicy, error) {
+	allPolicies, err := gc.svc.ListPolicies(ctx, &v1.RawQuery{})
+	if err != nil {
+		return []*storage.ListPolicy{}, errors.Wrap(err, "Failed to list policies from grpc client")
+	}
+
+	return allPolicies.Policies, err
+}
+
+func (gc *grpcClient) Get(ctx context.Context, id string) (*storage.Policy, error) {
+	policy, err := gc.svc.GetPolicy(ctx, &v1.ResourceByID{Id: id})
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to fetch policy %s", id)
+	}
+
+	return policy, nil
+}
+
+func (gc *grpcClient) Post(ctx context.Context, policy *storage.Policy) (*storage.Policy, error) {
+	req := &v1.PostPolicyRequest{
+		Policy:                 policy,
+		EnableStrictValidation: true,
+	}
+
+	policy, err := gc.svc.PostPolicy(ctx, req)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to POST policy")
+	}
+
+	return policy, nil
+}
+
+func (gc *grpcClient) Put(ctx context.Context, policy *storage.Policy) error {
+	_, err := gc.svc.PutPolicy(ctx, policy)
+
+	if err != nil {
+		return errors.Wrap(err, "Failed to PUT policy")
+	}
+
+	return nil
+}
+
+type client struct {
+	svc         PolicyClient
+	cache       map[string]*storage.Policy
+	lastUpdated time.Time
+}
+
+type clientOptions interface {
+	Apply(CachedPolicyClient)
+}
+
+func New(ctx context.Context, opts ...clientOptions) (CachedPolicyClient, error) {
+	c := client{}
+
+	for _, o := range opts {
+		o.Apply(&c)
+	}
+
+	if c.svc == nil {
+		gc, err := newGrpcClient(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not initialize policy client")
+		}
+
+		c.svc = gc
+	}
+
+	if err := c.FlushCache(ctx); err != nil {
+		return nil, errors.Wrap(err, "Failed to initialize cache")
+	}
+
+	return &c, nil
+}
+
+func (c *client) List(ctx context.Context) ([]*storage.Policy, error) {
+	list := make([]*storage.Policy, len(c.cache))
+	i := 0
+	for _, value := range c.cache {
+		list[i] = value
```

Comment:
Since we know up front how many policies are in the cache, I made a small optimization to initialize the array to the exact number of policies in the cache.  This means I have to assign the value like this.

On second look, it baffles me why I initialized my own `i` when I could have just used the first return value from `range`??  I'll try that out.

Reply:
Oh it's because `c.cache` is a map, not a list 😆

Reply:
policies := append(policies, value) will still work :-).  All you want is a list of policies in the cache (which happens to be keyed by name)

Reply:
I need to re-read https://go.dev/blog/slices to remember how this all works!

Reply:
ok so after reading that and looking at https://stackoverflow.com/a/38654841, looks like assigning to each array index is technically more efficient, but I do agree that `append()` is more readable and follows the typical pattern found in the code repo.  And efficiency is not of utmost concern here!  80fe7fa5d22c925c4ab10e1aa8f508bf79425827